README.md
# dvcz

**dvcz** is a
[distributed version control system](https://en.wikipedia.org/wiki/Distributed_version_control)
in the early stages of specification and development.  It can be seen as
a front end to
a group of loosely coordinated servers.

## Existing Code and Project Structure

dvcz is one of a number of projects using
[Github](https://www.github.com)
for backup.  Each of these has a **`.dvcz`** subdirectory and a number of
scripts in the project's root directory.

### Environmental Variables

These must be defined in order for project software to work correctly;
they include:

* DEV_BASE, the path to the directory containing project directories
* DVCZ_DIR, which might be for example `"/var/app/sharedev"`
* DVCZ_UDIR, where files are stored by content key; eg `"/var/app/sharedev/U"`
* DVCZ_USER_DIR, path to this user's .dvcz/, so `$HOME/.dvcz`
* DVCZ_AUTHOR, the double-quoted name of the author, such as `"John Smith"`
* DVCZ_AUTHOR_EMAIL, such as `"john.smith@example.com"`
* DVCZ_PATH_TO_KEYS, conventionally `"$DVCZ_USER_DIR/node"`

On a Linux system these are conventionally set in `$HOME/.bashrc`, a bash
script, with lines like

    export DEV_BASE="$HOME/dev"

### User .dvcz Subdirectory

This is `$DVCZ_USER_DIR` which is conventionally `$HOME/.dvcz`.  Files in or
below this subdirectory include

* `node/skPriv.pem`, ie `$DVCZ_PATH_TO_KEYS/skPriv.pem`, the RSA secret key
    used for signing BuildLists etc
* `id`, the globally unique 160-bit ID used to identify the user

### Project .dvcz Subdirectory

Files in this directory describe the current state of the project.  At
this point these include

* **builds**, a list of of timestamps, version numbers, and content hashes
* **lastBuildList**, the full text of the last build list, digitally signed
with the committer's RSA key
* **projCfg**, a description of what is to be included in each project build
* **projCfg.local**, a list of items (file or directory names)
    which must be included in `projCfg` if it is rewritten
* **version**, containing two newline-terminated lines, the first of which
    is a serialized `DecimalVersion` and the second a date in `CCYY-MM-DD`
    format

#### builds

This is a list in chronological order of

* the date and time at which each version of the project has been committed
* the corresponding version number
* and the content hash of the BuildList for the commit

An excerpt from a typical `builds` file:

    2016-05-19 18:31:36 v0.4.23 8a8ed84f8937a1f20c0d317c115e87fc9ebb12e8
    2016-05-25 19:53:11 v0.4.24 92d53f31e14c52de23830ad093db11440d77803a
    2016-05-25 20:04:30 v0.4.24 2026ccd5c10112e509108606f7770a4cd3f61a6c
    2016-06-15 00:32:54 v0.4.26 752e4eb5bca0033a6db4fb1391df120671c0daa9

Each line begins with a timestamp in `CCYY-MM-DD HH:MM:SS` format.
This is followed by a three or four part decimal version number, and
then the content key (the SHA hash of the contents of the BuildLists
committed).  The content key is 40 hexadecimal digits if using `SHA1`
and 64 hex digits if using `SHA2`.

### Bash Scripts

Currently each dvcz-assisted root project directory contains

* `bkp2U`, which backs the project up
* `countem`, which generates a rough number of lines of source code
* `styleit`, which styles code files appropriately for the language
* `test.all`, which runs language-specific tests

Python projects also have

* `checkCoverage`, which checks the degree to which tests cover the Python library
* `doPyLint`, which runs the utility `pylint` on Python language project files
* `installit`, which does a Python build, creating a tarball

#### bkp2U Script

This is a bash script in the project's root directory.  This is
`$DEV_BASE/$PROJECT` and so might resemble `/home/jdd/dev/dvcz`.

The `bkp2U` script backs up project files to the
content-keyed store, the path to which is set in `$DVCZ_UDIR`.   Before
doing the actual backup, the script changes the working directory to
the project's base directory,
`$DEV_BASE/$PROJECT`.  During the course of the backup the script generates a
BuildList for the contents of the project directory and subdirectories.
After committing
all of the project files, the script digitally signs the BuileList
using the RSA key
stored at `$DVCZ_PATH_TO_KEYS`.  This points to an RSA private key in
PEM format (and so might look like `$HOME/.dvcz/node`.  The key file
is conventionally named `skPriv.pem`.  That is, the path to the key
file is something like `/home/jdd/.dvcz/node/skPriv.pem`.

The signed build list is then (a) stored in `$DVCZ_UDIR` and (b) written
to `.dvcz/lastBuildList`.

`bkp2U` will eventually be replaced by `dvc_commit`.

### Utilities

#### dvc_adduser

This utility sets up a committer's `.dvcz` directory if necessary and then
adds a subdirectory to `uDir` for used by the committer in submitting files
to the content-keyed store (`uDir`).

For our purposes here, `.dvcz` must contain two files:

* `node/skPriv.pem`, which contains the committer's secret RSA key
* `id`, a 40- or 64-hex character sequence unique to the committer

If either of these is missing the utility will create it.

    usage: dvc_adduser [-h] [-f] [-j] [-k KEY_BITS] [-s DIR_STRUC_NAME] [-T] [-V]
                       [-1] [-2] [-3] [-u U_PATH] [-v]

    Set up directories for a new DVCZ user.

    optional arguments:
      -h, --help            show this help message and exit
      -f, --force           overwrite existing user configuration
      -j, --just_show       show options and exit
      -k KEY_BITS, --key_bits KEY_BITS
                            number of RSA key bits
      -s DIR_STRUC_NAME, --dir_struc_name DIR_STRUC_NAME
                            new dirStruc (DIR_FLAT, DIR16x16, or DIR256x256
      -T, --testing         this is a test run
      -V, --show_version    display version number and exit
      -1, --using_sha1      using the 160-bit SHA1 hash
      -2, --using_sha2      using the 256-bit SHA2 (SHA256) hash
      -3, --using_sha3      using the 256-bit SHA3 (Keccak-256) hash
      -B, --using_blake2b   using the blake2b hash with 256-bit digest
      -u U_PATH, --u_path U_PATH
                            path to uDir
      -v, --verbose         be chatty

#### dvc_commit

    usage: dvc_commit [-h] [-j] [-M MATCH_PAT] [-p PROJ_PATH] [-T] [-V] [-1] [-2]
                      [-3] [-u U_PATH] [-v] [-X EXCLUSIONS]

    generate build list for directory, optionally populating u_path

    optional arguments:
      -h, --help            show this help message and exit
      -j, --just_show       show options and exit
      -M MATCH_PAT, --match_pat MATCH_PAT
                            include only files matching this pattern
      -p PROJ_PATH, --proj_path PROJ_PATH
                            data directory for build list (default=../)
      -T, --testing         this is a test run
      -V, --showVersion     display version number and exit
      -1, --using_sha1      using the 160-bit SHA1 hash
      -2, --using_sha2      using the 256-bit SHA2 (SHA256) hash
      -3, --using_sha3      using the 256-bit SHA3 (Keccak-256) hash
      -B, --using_blake2b   using the blake2b hash with 256-bit digest
      -u U_PATH, --u_path U_PATH
                            path to uDir
      -v, --verbose         be chatty
      -X EXCLUSIONS, --exclusions EXCLUSIONS
                            do not include files/directories matching this pattern

### Testing

See
[DvcTestSetup](https://jddixon.github.io/dvcz/test_directory_structure.html)
for a description of how a DVCZ test directory is set up.

## Implementation Language

In its current form the dvcz library and scripts are written in Python 3.

## Dependencies

dvcz depends upon a number of related projects:

* [**buildlist**](https://jddixon.github.io/buildlist), a technology
for authenticating the contents of a directory structure
* [**fieldz**](https://jddixon.github.io/fieldz), a communications
prototol intended to be compatible with a subset of Google's
[Protocol Buffers](https://developers.google.com/protocol-buffers)
* **upax**, as described above

## Project Status

Pre-alpha.

In active development and currently used for backup but
generally not yet fully specified.

## On-line Documentation

More information on the **dvcz** project can be found
[here](https://jddixon.github.io/dvcz)


CHANGES
py/dvcz/CHANGES

v0.1.23
    2018-03-07
        * fix setup.py and tox config file                          SLOC 2414
v0.1.22
    2018-01-30
        * add skeletal blake2b support                              SLOC 2411
v0.1.21
    2017-12-09
        * pylint; minor bug fixes                                   SLOC 2413
v0.1.20
    2017-09-27
        * add projConfig.toml, correct version nbr in setup.py      SLOC 2418
v0.1.19
    2017-09-11
        * correct requirements.txt                                  SLOC 2390
v0.1.18
    2017-09-04
        * pep8 fixes; correct tox.ini                               SLOC 2380
v0.1.17
    2017-07-14
        * rewrite setup.py, MANIFEST.in; edit other config fieles   SLOC 2347
v0.1.16
    2017-06-09
        * restructure for pylint                                    SLOC 2369
v0.1.15
    2017-03-16
        * dvc_check_builds works                                    SLOC 2372
v0.1.14
    2017-03-09
        * skeletal dvcz_check_builds; update web pages              SLOC 2331
v0.1.13
    2017-02-27
        * update and improve docs                                   SLOC 2208
v0.1.12
    2017-02-25
        * actually add test_committer.py                            SLOC 2206
v0.1.11
    2017-02-04
        * correct size of RSA keys (off by 1); minor bug fixes      SLOC 1925
v0.1.10
    2017-01-22
        * improve specs in test_store.py, dvcz/store.py             SLOC 1919
v0.1.9
    2017-01-19
        * get DIR_FLAT etc from DirStruc; minor bug fixes           SLOC 1840
v0.1.8
    2017-01-12
        * modify to conform to revised xlattice_py API              SLOC 1821
v0.1.7
    2017-01-03
        * get rid of various minor errors, improve valid name tests SLOC 1816
v0.1.6
    2016-12-28
        * Committer, test_committer.py                              SLOC 1820
v0.1.5
    2016-12-27
        * add User, user.py, test_user.py                           SLOC 1664
v0.1.4
    2016-12-26
        * add Store, store.py, and tests                            SLOC 1565
v0.1.3
    2016-12-25
        * add Project, project.py, and tests                        SLOC 1498
v0.1.2
    2016-12-24
        * rearrange code, introducing user.py                       SLOC 1341
v0.1.1
    2016-12-23
        * add test_directory_structure.*                            SLOC 1341
v0.1.0
    2016-12-22
        * add {Pub,}{User,Committer} and Project abstractions       SLOC 1352
v0.0.20
    2016-12-21
        * skeletal test_dvc_setup.py                                SLOC 1216
v0.0.19
    2016-12-19
        * dvc_commit is noisy but works                             SLOC 1091
v0.0.18
    2016-12-18
        * reworking dvc_commit                                      SLOC 1047
v0.0.17
    2016-12-10
        * dvc_adduser creates $HOME/.dvcz                           SLOC 1173
v0.0.16
    2016-12-02
        * dvc_adduser stable, tested                                SLOC 1201
v0.0.15
    2016-12-01
        * enhanced dvc_adduser                                      SLOC 1189
v0.0.14
    2016-11-28
        * skeletal dvc_adduser                                      SLOC 1272
v0.0.13
    2016-11-27
        * mv dvc_get_project to library; PEP 8                      SLOC 1149
v0.0.12
    2016-11-24
        * add dvc_bkp2u                                             SLOC  950
v0.0.11
    2016-09-30
        * add pylint, coverage                                      SLOC  833
v0.0.10
    2016-09-17
        * add usingSHA = usingSHA{1,2,3}                            SLOC  802
v0.0.9
    2016-09-02
        * usingSHA1 -> usingSHA                                     SLOC  826
v0.0.8
    2016-08-23
        * edit gh-pages                                             SLOC  816
v0.0.7
    2016-08-12
        * elaborate project documentation                           SLOC  789
v0.0.6
    2016-06-21
        * add ghpDoc/{img,css}                                      SLOC  768
v0.0.5
    2016-05-18
        * add bkp2U, update .dvcz
v0.0.4
    2016-04-23
        * add .dvcz, mv .projCfg there as projCfg.local             SLOC  287
v0.0.3
    2016-04-21
        * add .projCfg, fix countem, edit dvcCommit                 SLOC  281
v0.0.2
    2016-04-18
        * add dvCommit for hacking on                               SLOC  208
v0.0.1
    2016-03-31
        * initial commit                                            SLOC   48








src/dvcz/store.py
# dvcz/store.py

"""
Content-keyed store for our distributed version control system.

In production serialized Store descriptors for a given user are collected under
    $HOME/.dvcz/stores/
That is, stores/ is a collection of files, where each file is a
serialized Store.  The name of the file is the name of the Store.
Such names are guaranteed to be unique within stores/.

The store itself is not serialized.  What is serialized is the Store,
whcih contains metadata about the store.
/
Each store is a UDir, a content-keyed store, and so looks like

    PATH_TO_STORE/
        in/
        tmp/
        --hash--
        --hash--
        ...

Depending on the directory structure and hash type, `--hash--` will
represent either a 4-bit/single hex digit directory name, an
8-bit/two hex digit directory name, or a 160-bit/40 hext digit
content key or a 256-bit/64 hex digit file name, where the file
name is the hash of the file contents.

`in/` and `tmp/` are scratch (temporary) UDirs used by the system.
Each user has a unique 256-bit/32-byte/64 hex digit userID.  A
commit copies user files into the store.  Files are initially
copied to a subdirectory of `in/` which is itself a UDir with a
DIR_FLAT directory structure.  So the commit operation creates

    in/
        USER_ID/        # a UDir, DIR_FLAT
            --hash--
            --hash--
            --hash--
            ...
            --hash--

If N files are being committed, N new entries will appear below
`in/USER_ID` when the operation is complete.`:

"""

# import hashlib

# from buildlist import(check_dirs_in_path, generate_rsa_key,
#                      read_rsa_key, rm_f_dir_contents)
from dvcz import DvczError
from dvcz.project import Project
from xlattice import HashTypes
from xlu import UDir, DirStruc

# from Crypto.PublicKey import RSA

# if sys.version_info < (3, 6):
#    import sha3

__all__ = ['Store']


class Store(UDir):
    """
    Link a name to a content-keyed store.

    The name should be unique within the context and must be a valid store
    name.  The name need have nothing to do with u_path.  Currently the
    rules for store name are the same as the rules for project names.

    If the directory at u_path already exists, its directory
    structure (dir_struc) and SHA hash type (hashtype) are discovered
    and override any attributes supplied.

    If u_path does not exist, the directory is created using the attributes
    passed.

    """

    def __init__(self, name, u_path, dir_struc=DirStruc.DIR_FLAT,
                 hashtype=HashTypes.SHA2, mode=0o755):

        if not Project.valid_proj_name(name):
            raise DvczError("not a valid store name: '%s'" % name)
        if not isinstance(dir_struc, DirStruc):
            raise DvczError("not a valid dir_struc: '%s'" % dir_struc)
        super().__init__(u_path, dir_struc, hashtype, mode)
        self._name = name

    @property
    def name(self):
        """
        Return the name assigned to the store.

        This is a convenience.  It is a valid store name, a single word
        incorporating no delimiters.
        """
        return self._name

    def __str__(self):
        return '::'.join([self.name,
                          self.u_path,
                          # pylint: disable=no-member
                          self.dir_struc.name,
                          self.hashtype.name])

    @classmethod
    def create_from_file(cls, path):
        """ Given an on-disk serialization, create a Store object. """

        with open(path, 'r') as file:
            # Possible FileNotFound.
            text = file.read()
        return cls.create_from_string(text)

    @classmethod
    def create_from_string(cls, text):
        """ Given a simple string serialization, create a Store object. """
        parts = text.split('::')
        pcount = len(parts)
        if pcount == 4:
            name = parts[0]
            u_path = parts[1]
            ds_name = parts[2]
            dir_struc = None
            for _ in DirStruc:
                if _.name == ds_name:
                    dir_struc = _
                    break
            else:
                raise DvczError(
                    "Not the name of a valid dir_struc name: '%s'" % ds_name)

            # 'item access'
            hashtype = HashTypes[parts[3]]
            return Store(name, u_path, dir_struc, hashtype)
        else:
            raise DvczError("Invalid Store descriptor: '%s'" % text)
